File: backend/app/models/okr.py
from datetime import datetime, timezone

from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from app.core.database import Base


class Objective(Base):
    __tablename__ = "objectives"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    title = Column(String(200), nullable=False)
    description = Column(Text, nullable=True)
    quarter = Column(String(10), nullable=False)  # e.g., "2025Q1"
    color = Column(String(20), default="blue")
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    key_results = relationship("KeyResult", backref="objective", cascade="all, delete-orphan")


class KeyResult(Base):
    __tablename__ = "key_results"

    id = Column(Integer, primary_key=True, index=True)
    objective_id = Column(Integer, ForeignKey("objectives.id"), nullable=False, index=True)
    title = Column(String(200), nullable=False)
    target_value = Column(Float, default=100.0)
    current_value = Column(Float, default=0.0)
    unit = Column(String(20), default="%")
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

File: frontend/src/api/okr.ts
import api from "./client";
import type { Objective, KeyResult } from "../types";

export const okrApi = {
  listObjectives: () =>
    api.get<Objective[]>("/okr/objectives").then((r) => r.data),

  createObjective: (data: {
    title: string;
    description?: string;
    quarter: string;
    color?: string;
  }) => api.post<Objective>("/okr/objectives", data).then((r) => r.data),

  updateObjective: (
    id: number,
    data: Partial<{ title: string; description: string; quarter: string; color: string }>
  ) => api.patch<Objective>(`/okr/objectives/${id}`, data).then((r) => r.data),

  deleteObjective: (id: number) => api.delete(`/okr/objectives/${id}`),

  createKeyResult: (
    objectiveId: number,
    data: { title: string; target_value?: number; current_value?: number; unit?: string }
  ) =>
    api
      .post<KeyResult>(`/okr/objectives/${objectiveId}/key-results`, data)
      .then((r) => r.data),

  updateKeyResult: (
    id: number,
    data: Partial<{ title: string; target_value: number; current_value: number; unit: string }>
  ) => api.patch<KeyResult>(`/okr/key-results/${id}`, data).then((r) => r.data),

  deleteKeyResult: (id: number) => api.delete(`/okr/key-results/${id}`),
};
